mod-clans/src/main/java/gg/solarmc/loader/clans/Clan.java

```diff
@@ -0,0 +1,407 @@
+                .values(this.clanID,receiver.getUserId())
+                .execute();
+
+        if (sec != 1) throw new IllegalStateException("Inserted gg.solarmc.loader.clans.ClanMember already belongs to clan!");
+
+        receiver.setCachedClan(this);
+
+        return getClanMembers(transaction);
+    }
+
+    /**
+     * Removes a member from the clan.
+     * @param transaction The tx
+     * @param receiver gg.solarmc.loader.clans.ClanDataObject to remove
+     * @return Accurate set of ClanMembers post transaction
+     * @throws IllegalStateException if member already has a clan
+     * @throws IllegalArgumentException if member tried to remove was the leader of the clan, or if they are not
+     * present in the clan
+     */
+    public Set<ClanMember> removeClanMember(Transaction transaction, ClanDataObject receiver) {
+        if (receiver.getClan(transaction).isPresent()) throw new IllegalStateException("Receiver already has clan!");
+
+        ClansClanMembershipRecord rec = transaction.getProperty(DSLContext.class)
+                .fetchOne(CLANS_CLAN_MEMBERSHIP,CLANS_CLAN_MEMBERSHIP.CLAN_ID.eq(this.clanID).and(CLANS_CLAN_MEMBERSHIP.USER_ID.eq(receiver.getUserId())));
+
+        if (rec == null) throw new IllegalArgumentException("Member is not part of clan!");
+        if (receiver.isSimilar(leader)) throw new IllegalArgumentException("Tried to remove leader of clan!");
+
+        rec.delete();
+
+        receiver.setCachedClan(null);
+
+        return getClanMembers(transaction);
+    }
+
+    /**
+     * Adds clan as an ally
+     * @param transaction the tx
+     * @param receiver the receiver
+     * @return whether record was inserted or not.
+     * @throws IllegalArgumentException if receiver is the same gg.solarmc.loader.clans.Clan as this object
+     * @throws IllegalStateException if receiver or this object already has an ally
+     */
+    public boolean addClanAsAlly(Transaction transaction, Clan receiver) {
+        if (receiver.getID() == this.clanID) throw new IllegalArgumentException("Tried to mark clan ally as same clan!");
+        if (this.getAlliedClan(transaction).isPresent()) throw new IllegalStateException("Sender already have an ally!");
+        if (receiver.getAlliedClan(transaction).isPresent()) throw new IllegalStateException("Selected clan already has an ally!");
+
+        int res = transaction.getProperty(DSLContext.class)
+                .insertInto(CLANS_CLAN_ALLIANCES)
+                .columns(CLANS_CLAN_ALLIANCES.CLAN_ID,CLANS_CLAN_ALLIANCES.ALLY_ID)
+                .values(this.clanID,receiver.getID())
+                .execute();
+
+        this.alliedClan = receiver.getID();
+        receiver.alliedClan = this.getID();
+
+        return res == 1;
+    }
+
+    /**
+     * Revokes ally. Can be called from either clan. Revokes alliance for both.
+     * @param transaction the tx
+     * @return whether operation was successful or not
```

Comment:
```suggestion
     * @return true if the ally was removed, false if this clan did not have an ally
```

I wonder whether it would be better to return an Optional<Clan> which returns the Clan that was removed as an ally, or an empty optional if there was no ally.

Reply:
Done. I don't think returning an optional would have a point since what would you use it for? ( Same argument about returning a list of kits after removing a kit)